viper_test.go

```diff
@@ -2339,44 +2341,182 @@ func TestConfigChangedAt(t *testing.T) {
+			expectErr: true,
+		},
+		{
+			t:         time.Now(),
+			val:       "not a time",
+			expected:  time.Time{},
+			expectErr: true,
+		},
+	} {
+		t.Run(fmt.Sprintf("case=%d with value %+v", i, tc.val), func(t *testing.T) {
+			res, err := castAllSourcesE(tc.t, tc.val)
+			assert.Equal(t, tc.expectErr, err != nil)
+			assert.Equal(t, tc.expected, res)
+		})
+	}
+}
+
+func TestCastStringSourcesE(t *testing.T) {
+	ignore := "ignore"
+	for i, tc := range []struct {
+		t         interface{}
+		val       string
+		pflagVal  string
+		expected  interface{}
+		expectErr bool
+	}{
+		{
+			t:         []string{},
+			val:       "bar",
+			pflagVal:  "[bar]",
+			expected:  []string{"bar"},
+			expectErr: false,
+		},
+		{
+			t:         []string{},
+			val:       "a b c",
+			pflagVal:  "[a,b,c]",
+			expected:  []string{"a", "b", "c"},
+			expectErr: false,
+		},
+		{
+			t:         []int{},
+			val:       ignore, // this testcase is not possible as the cast library does not convert from string to []int
+			pflagVal:  "[1,2,3]",
+			expected:  []int{1, 2, 3},
+			expectErr: false,
+		},
+		{
+			t:         false,
+			val:       "true",
+			pflagVal:  "true",
+			expected:  true,
+			expectErr: false,
+		},
+		{
+			t:         0,
+			val:       "1234",
+			pflagVal:  "1234",
+			expected:  1234,
+			expectErr: false,
+		},
+		{
+			t:         0.0,
+			val:       "12.34",
+			pflagVal:  "12.34",
+			expected:  12.34,
+			expectErr: false,
+		},
+		{
+			t:         time.Nanosecond,
+			val:       "1s",
+			pflagVal:  "1s",
+			expected:  time.Second,
+			expectErr: false,
+		},
+		{
+			t:         time.Time{},
+			val:       time.Now().Add(time.Hour).Truncate(time.Hour * 24).UTC().String(),
+			pflagVal:  time.Now().Add(time.Hour).Truncate(time.Hour * 24).UTC().String(),
+			expected:  time.Now().Add(time.Hour).Truncate(time.Hour * 24).UTC(),
+			expectErr: false,
+		},
+		{
+			t:         0,
+			val:       "not a number",
+			pflagVal:  "not a number",
+			expected:  0,
+			expectErr: true,
+		},
+	} {
+		t.Run(fmt.Sprintf("case=%d value:%s pflagValue:%s", i, tc.val, tc.pflagVal), func(t *testing.T) {
+			if tc.val != ignore {
+				res, err := castStringSourcesE(tc.t, tc.val, false)
+				assert.Equal(t, tc.expectErr, err != nil, err)
+				assert.Equal(t, tc.expected, res)
+			}
 
-		key := "foo"
-		SetType(key, []string{})
-		Set(key, "bar")
-		assert.Equal(t, []string{"bar"}, Get(key))
+			if tc.pflagVal != ignore {
+				res, err := castStringSourcesE(tc.t, tc.pflagVal, true)
+				assert.Equal(t, tc.expectErr, err != nil, err)
+				assert.Equal(t, tc.expected, res)
+			}
+		})
+	}
+}
+
+func TestGetType(t *testing.T) {
+	t.Run("case=SetType has precedence over default", func(t *testing.T) {
+		v := New()
+
+		key := "test_key"
+		typ := ""
+		v.SetTypeByDefaultValue(true)
+		v.SetDefault(key, 0)
+		v.SetType(key, typ)
+		assert.Equal(t, typ, v.getType(key))
 	})
 
-	t.Run("case=converts string to string slice length 3", func(t *testing.T) {
```

Comment:
Should these really be removed? I think the new test cases should be additional to these

Reply:
they are restructured here https://github.com/ory/viper/pull/16/files#diff-db3e9f98aa59a770129eb45f83357512R2344